flang/lib/Lower/CallInterface.cpp

```diff
@@ -240,23 +240,27 @@ void Fortran::lower::CallerInterface::walkResultExtents(
   // descriptor inquiries to it that would fail to lower on the caller side).
   const Fortran::semantics::Symbol *interfaceSymbol =
       procRef.proc().GetInterfaceSymbol();
-  assert(interfaceSymbol &&
-         "can only walk result extent of user procedure with interface");
-  const Fortran::semantics::Symbol &result =
-      interfaceSymbol->get<Fortran::semantics::SubprogramDetails>().result();
-  if (const auto *objectDetails =
-          result.detailsIf<Fortran::semantics::ObjectEntityDetails>())
-    if (objectDetails->shape().IsExplicitShape())
-      for (const Fortran::semantics::ShapeSpec &shapeSpec :
-           objectDetails->shape())
-        visitor(Fortran::evaluate::AsGenericExpr(getExtentExpr(shapeSpec)));
+  if (interfaceSymbol) {
+    const Fortran::semantics::Symbol &result =
+        interfaceSymbol->get<Fortran::semantics::SubprogramDetails>().result();
+    if (const auto *objectDetails =
+            result.detailsIf<Fortran::semantics::ObjectEntityDetails>())
+      if (objectDetails->shape().IsExplicitShape())
+        for (const Fortran::semantics::ShapeSpec &shapeSpec :
+             objectDetails->shape())
+          visitor(Fortran::evaluate::AsGenericExpr(getExtentExpr(shapeSpec)));
+  } else {
+    assert(procRef.Rank() == 0 &&
+           "only scalar functions may not have an interface symbol");
```

Comment:
Looks like we could even put a fir::emitFatalError here, no? If we get to this condition, doesn't that mean semantics checking failed somehow?

Reply:
Yes, I think `procRef.Rank() != 0 && procRef.proc().GetInterfaceSymbol() != nullptr` is a predicate entirely enforced by the front-end. I made it a `fir::emitFatalError`.